src/muc/mod_muc_room.erl

```diff
@@ -900,7 +900,12 @@ can_send_broadcasts(Role, StateData) ->
 
 broadcast_room_packet(From, FromNick, Role, Packet, StateData) ->
     Activity = get_user_activity(From, StateData),
-    TS = Activity#activity.message_time,
+    TS = case Activity#activity.message_time of
+            0 ->
```

Comment:
When does this equal zero? What are the conditions for that?

Reply:
always. Maybe message_time is set when the user is shaped and waiting.
Code in MUC was weird, so probably that is why we always use 0 and mongoose_mam_id returns `max(previousMamId, 0) = previousMamId`, where `previousMamId` is set to a big proper value by mod_mam_pm.

```erlang
get_user_activity(JID, StateData) ->
    case treap:lookup(jid:to_lower(JID), StateData#state.activity) of
    {ok, _P, A} -> A; %% Try to search, but we will not hit this branch is we do not shape
    error ->
        %% XXX timestamp is missing, we make a new #activity{} with default timestamp=0
        MessageShaper = mongoose_shaper:new(get_opt(StateData, user_message_shaper)),
        PresenceShaper = mongoose_shaper:new(get_opt(StateData, user_presence_shaper)),
        #activity{message_shaper = MessageShaper,
                  presence_shaper = PresenceShaper}
    end.
    
store_user_activity(JID, UserActivity, StateData) -> %% where UserActivity = get_user_activity(JID, StateData)
...
    StateData1 =
    %% IF has no shapers
    case (MinMessageInterval == 0) andalso
        (MinPresenceInterval == 0) andalso
        (UserActivity#activity.message_shaper == none) andalso
        (UserActivity#activity.presence_shaper == none) andalso
        (UserActivity#activity.message == undefined) andalso
        (UserActivity#activity.presence == undefined) of
        true ->
        %% Literally do nothing here :(
        StateData#state{activity = Activity};
        false ->
        %% Store activity, but only if we have shapers
        %%  ....
            StateData#state{
              activity = treap:insert(
                       Key,
                       Priority,
                       UserActivity,
                       Activity)}
        end
    end,
    StateData1.
```

I could check older versions of code to check if get_user_activity(JID, StateData) ever set timestamp

Reply:
Here is how it looks like:

store_user_activity is called with `activity={activity,0,1717509341936627,none,none,undefined,undefined}  and jid="{jid,<<\"alice_unnamed_176\">>,<<\"localhost\">>,<<\"res1\">>}"`
It returns  `State` with `#state.activity=nil`

After we call  `get_user_activity` with passing `State.#state.activity=nil` to it from this nested function stack:
```
filter_room_packet from broadcast_room_packet
broadcast_room_packet from process_message_from_allowed_user
process_message_from_allowed_user from process_groupchat_message
process_groupchat_message from route_message
```